File: src/5/5.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"slices"
	"strconv"
	"strings"
)

type Range struct {
	min  int64
	max  int64
	diff int64
}

func main() {
	input := bufio.NewScanner(os.Stdin)
	br := false
	ranges := make([]Range, 200)
	ids := make([]int64, 1000)
	fmt.Println("Go!")
	for input.Scan() {
		line := input.Text()
		if line == "" && !br {
			br = true
			slices.SortFunc(ranges, func(a, b Range) int {
				min := a.min - b.min
				if min != 0 {
					return int(min)
				}
				return int(a.max - b.max)
			})
		} else if line == "" && br {
			break
		} else if !br {
			splitted := strings.Split(line, "-")
			min, _ := strconv.ParseInt(splitted[0], 0, 64)
			max, _ := strconv.ParseInt(splitted[1], 0, 64)
			ranges = append(ranges, Range{min, max, max - min + 1})
		} else if br {
			id, _ := strconv.ParseInt(line, 10, 64)
			ids = append(ids, id)
		}
	}
	suma := 0
	for _, id := range ids {
		for _, rang := range ranges {
			if rang.max-id <= rang.diff && rang.max-id > 0 {
				suma++
				break
			}
		}
	}
	sumb := int64(0)
	i := int64(0)
	for _, rang := range ranges {
		if rang.max >= i {
			sumb += (rang.max - max(rang.min, i)) + 1
			i = rang.max + 1
		}
	}
	fmt.Println(suma, sumb-1)
}

File: src/05/5.js
const readline = require('readline');

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout,
    terminal: false
});

let ranges = [];
let ids = [];
let br = false;

const differ = (upper, lower) => {
    return upper - lower + 1
}

rl.on('line', (line) => {
    if (line == '') {
        br = true;
    } else if (!br) {
        let nums = line.split('-').map(Number);
        ranges.push({ upper: nums[1], lower: nums[0], diff: differ(nums[1], nums[0]) });
        ranges.sort((a, b) => {
            let min = a.lower - b.lower;
            if (min != 0) {
                return min;
            }
            return a.upper - b.upper
        })
    } else if (br) {
        ids.push(Number(line));
    }
});

rl.on('close', () => {

    console.log('Go!');
    let suma = 0;
    for (let id of ids) {
        for (let range of ranges) {
            if (range.upper - id <= range.diff && range.upper - id > 0) {
                suma++;
                break;
            }
        }
    }
    let sumb = 0
    let i = 0
    for (let r of ranges) {
        if (r.upper >= i) {
            sumb += r.upper - Math.max(r.lower, i) + 1
            i = r.upper + 1
        }
    }
    console.log(suma, sumb)
})



